core/node/state_keeper/src/testonly/test_batch_executor.rs

```diff
@@ -805,55 +803,31 @@ impl StateKeeperIO for TestIO {
     }
 }
 
-/// `BatchExecutor` which doesn't check anything at all. Accepts all transactions.
-// FIXME: move to `utils`?
+/// Storage factory that produces empty VM storage for any batch. Should only be used with a mock batch executor
+/// that doesn't read from the storage. Prefer using `ConnectionPool` as a factory if it's available.
 #[derive(Debug)]
-pub(crate) struct MockBatchExecutor;
+pub struct MockReadStorageFactory(tempfile::TempDir);
 
-#[async_trait]
-impl BatchExecutor for MockBatchExecutor {
-    async fn init_batch(
-        &mut self,
-        _storage_factory: Arc<dyn ReadStorageFactory>,
-        _l1batch_params: L1BatchEnv,
-        _system_env: SystemEnv,
-        _stop_receiver: &watch::Receiver<bool>,
-    ) -> Option<BatchExecutorHandle> {
-        let (send, recv) = mpsc::channel(1);
-        let handle = tokio::task::spawn(async {
-            let mut recv = recv;
-            while let Some(cmd) = recv.recv().await {
-                match cmd {
-                    Command::ExecuteTx(_, resp) => resp.send(successful_exec()).unwrap(),
-                    Command::StartNextL2Block(_, resp) => resp.send(()).unwrap(),
-                    Command::RollbackLastTx(_) => panic!("unexpected rollback"),
-                    Command::FinishBatch(resp) => {
-                        // Blanket result, it doesn't really matter.
-                        resp.send(default_vm_batch_result()).unwrap();
-                        break;
-                    }
-                    Command::FinishBatchWithCache(resp) => resp
-                        .send((default_vm_batch_result(), storage_view_cache()))
-                        .unwrap(),
-                }
-            }
-            anyhow::Ok(())
-        });
-        Some(BatchExecutorHandle::from_raw(handle, send))
+impl Default for MockReadStorageFactory {
+    fn default() -> Self {
+        Self(
+            tempfile::TempDir::new()
+                .expect("failed creating temporary directory for `MockReadStorageFactory`"),
+        )
     }
 }
 
-#[derive(Debug)]
-pub struct MockReadStorageFactory;
-
 #[async_trait]
 impl ReadStorageFactory for MockReadStorageFactory {
     async fn access_storage(
         &self,
-        _stop_receiver: &Receiver<bool>,
+        _stop_receiver: &watch::Receiver<bool>,
         _l1_batch_number: L1BatchNumber,
-    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'_>>> {
-        // Presume that the storage is never accessed in mocked environment
-        unimplemented!()
+    ) -> anyhow::Result<Option<OwnedStorage>> {
+        let storage = RocksdbStorage::builder(self.0.path())
+            .await
+            .expect("Cannot create mock RocksDB storage")
+            .build_unchecked();
+        Ok(Some(PgOrRocksdbStorage::Rocksdb(storage).into()))
     }
```

Comment:
This is quite ugly, but I haven't found a way to mock the storage factory otherwise; with the changes, `access_storage` *is* called by the state keeper regardless of the `BatchExecutor` implementation. I guess theoretically it could be possible to add a mock storage variant to `PgOrRocksdbStorage`, but that would introduce unnecessary noise and may marginally hurt performance.

Reply:
It would be cool if we could have an associated type in `ReadStorageFactory` that would regulate which storage type is returned from a particular implementation. Don't think it's feasible though due to our heavy usage of `dyn`

Reply:
Yes, if we could have a varying output type, there wouldn't be a problem, but as you've said, it doesn't look compatible with trait objects. Will think about it a bit more.